odl/discr/discr_ops.py

```diff
@@ -130,110 +144,212 @@ def finite_diff(f, out=None, axis=0, dx=1.0, edge_order=2,
+            slice_node1[axis] = 1
+            slice_node2[axis] = 0
+            # 1D equivalent: out[0] = f[1] - f[0]
+            out[slice_out] = f_arr[slice_node1] - f_arr[slice_node2]
+
+            # 1st-oder upper edge
+            slice_out[axis] = -1
+            # 1D equivalent: out[-1] = f[-1] - f[-1] = 0
+            out[slice_out] = 0
+
+        elif method == 'backward':
+            # 1st-oder lower edge
+            slice_out[axis] = 0
+            # 1D equivalent: out[0] = f[0] - f[0] = 0
+            out[slice_out] = 0
+
+            # 1st-oder upper edge
+            slice_out[axis] = -1
+            slice_node1[axis] = -1
+            slice_node2[axis] = -2
+            # 1D equivalent: out[-1] = f[-1] - f[-2]
+            out[slice_out] = f_arr[slice_node1] - f_arr[slice_node2]
+
+    # Use one-sided differences on the endpoints
     else:
 
-        # Numerical differentiation: 1st order edges
-        if f_data.shape[axis] == 2 or edge_order == 1:
+        # Edge-order accuracy is triggered implicitly by the method used in
+        #  the interior or explicitly using ``edge_order``
+
+        # 1st order edges
+        if edge_order == 1:
+            # lower boundary
             slice_out[axis] = 0
             slice_node1[axis] = 1
             slice_node2[axis] = 0
             # 1D equivalent: out[0] = (f[1] - f[0])
-            out[slice_out] = f_data[slice_node1] - f_data[slice_node2]
+            out[slice_out] = f_arr[slice_node1] - f_arr[slice_node2]
 
+            # upper boundary
             slice_out[axis] = -1
             slice_node1[axis] = -1
             slice_node2[axis] = -2
             # 1D equivalent: out[-1] = (f[-1] - f[-2])
-            out[slice_out] = f_data[slice_node1] - f_data[slice_node2]
+            out[slice_out] = f_arr[slice_node1] - f_arr[slice_node2]
 
-        # Numerical differentiation: 2nd order edges
-        else:
+        # 2nd order edges
+        elif edge_order == 2:
+            # lower boundary
             slice_out[axis] = 0
             slice_node1[axis] = 0
             slice_node2[axis] = 1
             slice_node3[axis] = 2
             # 1D equivalent: out[0] = -(3*f[0] - 4*f[1] + f[2]) / 2.0
-            out[slice_out] = -(3.0 * f_data[slice_node1] - 4.0 * f_data[
-                slice_node2] + f_data[slice_node3]) / 2.0
+            out[slice_out] = -(3.0 * f_arr[slice_node1] - 4.0 * f_arr[
+                slice_node2] + f_arr[slice_node3]) / 2.0
 
+            # upper boundary
             slice_out[axis] = -1
             slice_node1[axis] = -1
             slice_node2[axis] = -2
             slice_node3[axis] = -3
             # 1D equivalent: out[-1] = (3*f[-1] - 4*f[-2] + f[-3]) / 2.0
-            out[slice_out] = (3.0 * f_data[slice_node1] - 4.0 * f_data[
-                slice_node2] + f_data[slice_node3]) / 2.0
+            out[slice_out] = (3.0 * f_arr[slice_node1] - 4.0 * f_arr[
+                slice_node2] + f_arr[slice_node3]) / 2.0
```

Comment:
This function has become really monstrously long. You could put the code for the computation of the values of the slices a small helper which takes the sliced arrays and the right hand sides and just sets according to those parameters?